system/metrics/sinks/file/fileSink.cpp

```diff
@@ -25,82 +25,38 @@ extern "C" MetricSink* getSinkInstance(const char *name, const IPropertyTree *pS
 
 
 FileMetricSink::FileMetricSink(const char *name, const IPropertyTree *pSettingsTree) :
-        MetricSink(name, "file"),
-        collectionPeriodSeconds{60}
+    PeriodicMetricSink(name, "file", pSettingsTree)
 {
-    if (pSettingsTree->hasProp("@period"))
-    {
-        collectionPeriodSeconds = pSettingsTree->getPropInt("@period");
-    }
-
     pSettingsTree->getProp("@filename", fileName);
     clearFileOnStartCollecting = pSettingsTree->getPropBool("@clear", false);
 }
 
 
-FileMetricSink::~FileMetricSink()
-{
-    if (isCollecting)
-    {
-        doStopCollecting();
-    }
-}
-
-
-void FileMetricSink::startCollection(MetricsReporter *_pReporter)
+void FileMetricSink::prepareToStartCollecting()
 {
     fhandle = fopen(fileName.str(), clearFileOnStartCollecting ? "w" : "a");
-    pReporter = _pReporter;
-    isCollecting = true;
-    collectThread = std::thread(&FileMetricSink::collectionThread, this);
 }
 
-
-void FileMetricSink::collectionThread()
-{
-    //
-    // The initial wait for the first report
-    waitSem.wait(collectionPeriodSeconds * 1000);
-    while (!stopCollectionFlag)
-    {
-        auto reportMetrics = pReporter->queryMetricsForReport(name);
-        writeReportHeaderToFile();
-        for (auto &pMetric: reportMetrics)
-        {
-            writeMeasurementToFile(pMetric->queryName(), pMetric->queryValue(), pMetric->queryDescription());
-        }
-
-        // Wait again
-        waitSem.wait(collectionPeriodSeconds * 1000);
-    }
-}
-
-
-void FileMetricSink::stopCollection()
+void FileMetricSink::doCollection()
 {
-    if (isCollecting)
+    auto reportMetrics = pReporter->queryMetricsForReport(name);
+    writeReportHeaderToFile();
+    for (auto &pMetric: reportMetrics)
     {
-        doStopCollecting();
+        writeMeasurementToFile(pMetric->queryName(), pMetric->queryValue(), pMetric->queryDescription());
     }
 }
 
 
-void FileMetricSink::doStopCollecting()
+void FileMetricSink::collectingHasStopped()
 {
-    //
-    // Set the stop collecting flag, then signal the wait semaphore
-    // to wake up and stop the collection thread
-    stopCollectionFlag = true;
-    waitSem.signal();
-    isCollecting = false;
-    collectThread.join();
     fclose(fhandle);
 }
 
 
-void FileMetricSink::writeMeasurementToFile(const std::string &metricName, uint32_t value, const std::string &metricDescription) const
+void FileMetricSink::writeMeasurementToFile(const std::string &metricName, uint64_t value, const std::string &metricDescription) const
 {
-    fprintf(fhandle, "  %s -> %d, %s\n", metricName.c_str(), value, metricDescription.c_str());
+    fprintf(fhandle, "  %s -> %" PRIu64 ", %s\n", metricName.c_str(), value, metricDescription.c_str());
```

Comment:
Unfortunately PRIu64 is a C99 extension, and not part of c++ (it is a bit of a mess).